samza-core/src/main/java/org/apache/samza/storage/blobstore/BlobStoreBackupManager.java

```diff
@@ -0,0 +1,349 @@
+    storeSCMs.forEach((storeName, snapshotIndexBlobId) -> {
+      // Only perform cleanup for stores configured with BlobStore State Backend Factory
+      if (storesWithBlobStoreStateBackend.contains(storeName)) {
+        Metadata requestMetadata =
+            new Metadata(Metadata.SNAPSHOT_INDEX_PAYLOAD_PATH, Optional.empty(), jobName, jobId, taskName, storeName);
+        CompletionStage<SnapshotIndex> snapshotIndexFuture =
+            blobStoreUtil.getSnapshotIndex(snapshotIndexBlobId, requestMetadata);
+
+        // 1. remove TTL of index blob and all of its files and sub-dirs marked for retention
+        CompletionStage<Void> removeTTLFuture =
+            snapshotIndexFuture.thenComposeAsync(snapshotIndex -> {
+              LOG.debug("Removing TTL for index blob: {} and all of its files and sub-dirs for task: {} store :{}",
+                  snapshotIndexBlobId, taskName, storeName);
+              return blobStoreUtil.removeTTL(snapshotIndexBlobId, snapshotIndex, requestMetadata);
+            }, executor);
+        removeTTLFutures.add(removeTTLFuture);
+
+        // 2. delete the files/subdirs marked for deletion in the snapshot index.
+        CompletionStage<Void> cleanupRemoteSnapshotFuture =
+            snapshotIndexFuture.thenComposeAsync(snapshotIndex -> {
+              LOG.debug("Deleting files and dirs to remove for current index blob: {} for task: {} store: {}",
+                  snapshotIndexBlobId, taskName, storeName);
+              return blobStoreUtil.cleanUpDir(snapshotIndex.getDirIndex(), requestMetadata);
+            }, executor);
+
+        cleanupRemoteSnapshotFutures.add(cleanupRemoteSnapshotFuture);
+
+        // 3. delete the remote {@link SnapshotIndex} blob for the previous checkpoint.
+        CompletionStage<Void> removePrevRemoteSnapshotFuture =
+            snapshotIndexFuture.thenComposeAsync(snapshotIndex -> {
+              if (snapshotIndex.getPrevSnapshotIndexBlobId().isPresent()) {
+                String blobId = snapshotIndex.getPrevSnapshotIndexBlobId().get();
+                LOG.debug("Removing previous snapshot index blob: {} from blob store for task: {} store: {}.",
+                    blobId, taskName, storeName);
+                return blobStoreUtil.deleteSnapshotIndexBlob(blobId, requestMetadata);
+              } else {
+                // complete future immediately. There are no previous snapshots index blobs to delete.
+                return CompletableFuture.completedFuture(null);
+              }
+            }, executor);
+        removePrevRemoteSnapshotFutures.add(removePrevRemoteSnapshotFuture);
+      }
+    });
+
+    return FutureUtil.allOf(removeTTLFutures, cleanupRemoteSnapshotFutures, removePrevRemoteSnapshotFutures)
+        .whenComplete((res, ex) -> metrics.cleanupNs.update(System.nanoTime() - startTime));
+  }
+
+  @Override
+  public void close() {
+    blobStoreUtil.closeBlobStoreManager();
```

Comment:
Why is BlobStoreUtil proxying the close call to the Manager?

Reply:
Fixed that to call init/close directly.